### PR TITLE
Add progress meter to `run!`

### LINKED ISSUE #542: Progress meter
I vaguely remember there used to be an option to print the progress of a simulation in the `run!` function. Am I correct? Any reason it has been taken out? This is an important feature for heavy simulations.

### PR DESCRIPTION
**Pull request by kavir1698**
Closes #542

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 name = "Agents"
 uuid = "46ada45e-f475-11e8-01d0-f70cc89e6671"
 authors = ["George Datseris", "Tim DuBois", "Aayush Sabharwal", "Ali Vahdati"]
-version = "5.3.0"
+version = "5.3.1"
```
Comment (Datseris) [inline Project.toml:4]: This should be 5.4.0 because it adds a new feature, it does not fix a bug.

File: src/simulations/collect.jl [hunk 2]
```diff
@@ -112,12 +113,13 @@ function run!(
     agent_step!,
     model_step!,
     n;
-    when = true,
-    when_model = when,
-    mdata = nothing,
-    adata = nothing,
-    obtainer = identity,
-    agents_first = true,
+    when=true,
```
Comment (Datseris) [inline src/simulations/collect.jl:116]: I don't understand these changes. It changes the more readable format of `x = y` to the less readable one of `x=y`. Regardless, even though this style can be debated of which one is better, such formatting style changes shouldn't be done in a PR that adds a new feature, but in an independent one.

File: src/simulations/collect.jl [hunk 3]
```diff
@@ -145,6 +148,7 @@ function run!(
         end
         step!(model, agent_step!, model_step!, 1, agents_first)
         s += 1
+        next!(p)
```
Comment (Datseris) [inline src/simulations/collect.jl:151]: Do `showprogress && next!(p)` here.

File: src/simulations/collect.jl [hunk 4]
```diff
@@ -136,6 +138,7 @@ function run!(
     end
 
     s = 0
+    p = showprogress ? typeof(n) <: Int ? ProgressMeter.Progress(n, 1) : [0] : nothing
```
Comment (Datseris) [inline src/simulations/collect.jl:141]: This is not the correct way to use the `ProgressMeter` package. Instead you need to initialize a proper progress meter as indicated by the package's API. `p = ProgressMeter.Progress(n, 0, "Performing `Agents.run!`...")` and later  down the line call `next!(p)`.

Comment (Datseris) [reply src/simulations/collect.jl:141]: Furthermore, the ProgressMeter.jl package has a `ProgressUnknown` type that we can use when `n` is a function: https://github.com/timholy/ProgressMeter.jl#progress-meters-for-tasks-with-an-unknown-number-of-steps

File: src/simulations/collect.jl [hunk 5]
```diff
@@ -510,3 +514,20 @@ collect_model_data!(df, model, properties::Nothing, step::Int = 0; kwargs...) =
...
+
+
+###################################################
+# ProgressMeter 
+###################################################
+
+"""
+Show number of passed steps when `n` in `run!` is a function.
+"""
+function next!(p::Vector{Int})
```
Comment (Datseris) [inline src/simulations/collect.jl:526]: This is type piracy and definitely not something we can do in a Julia package. It extends a method whose both function, and type of input arguments, we do not own. The ProgressMeter.jl package says how we should use it by initializing a `Progress`.

**Review by Datseris — CHANGES_REQUESTED** [review]
Adding progress meter report to `run!` is a great idea, but at the moment the PR uses ProgressMeter.jl inappropriately. Once the documentation of ProgressMeter.jl has been consulted and the PR has been modified to use the expected API, I'm happy to merge!

File: src/simulations/collect.jl (hunk 2 above)
Comment (kavir1698) [reply src/simulations/collect.jl]: This was something Julia addon on VSCode changed automatically. I reverted them.

File: src/simulations/collect.jl (hunk 4 above)
Comment (kavir1698) [reply src/simulations/collect.jl]: Thanks for your feedback. I wasn't aware of `ProgressUnknown`.

**Review by Datseris — APPROVED** [review]
Awesome! I didn't know that you could set visibility from within the Progress type.

**Comment by Datseris** [comment]
@kavir1698 forgot to ask: did you test this also with parallel runs and it works? And what happens when we do ensemvle runs? Does it show several progress bars in paralell?

**Comment by kavir1698** [comment]
> @kavir1698 forgot to ask: did you test this also with parallel runs and it works? And what happens when we do ensemvle runs? Does it show several progress bars in paralell?

I just checked it on `paramscan`. It prints one progress bar for each simulation. The best solution is to `showprogress` to `false` by default? so that no other change to code is needed.

**Comment by Datseris** [comment]
yeah, then I assume on `ensemblerun` it would do the same, one progressbar per simulation. So it is probably better to have it false by default. Then, in ensemble run, we should have a progressbar that counts the full simulations, not steps.

**Pull request merged**